File: tools/glosh-remote-spike/app/src/main/java/com/glosh/remote/spike/guide/autopilot/AdaptiveAutopilotPlanner.java
package com.glosh.remote.spike.guide.autopilot;

import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Action;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Decision;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Observation;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Screen;

public final class AdaptiveAutopilotPlanner {
    public static final int MIN_WIRELESS_ADB_API = 30;

    public Decision decide(Observation observation) {
        if (observation.supportConnected()) {
            return decision(Action.DONE, "support already connected");
        }
        if (observation.adbConnected()) {
            return decision(Action.CONNECT_SUPPORT, "ADB already valid; skip Settings/pairing");
        }
        if (observation.androidApi() < MIN_WIRELESS_ADB_API) {
            return decision(Action.UNSUPPORTED_ANDROID, "Android <11 standard wireless path unsupported");
        }
        if (!observation.accessibilityEnabled() && observation.restrictedSettingsRequired()) {
            return decision(Action.ASK_ALLOW_RESTRICTED_SETTINGS, "sideloaded Accessibility is restricted by OS");
        }
        if (!observation.accessibilityEnabled()) {
            return decision(Action.ASK_ENABLE_ACCESSIBILITY, "manual bootstrap prerequisite");
        }
        if (!observation.wifiReady()) {
            return decision(Action.ASK_CONNECT_WIFI, "Wireless Debugging needs a usable Wi-Fi network");
        }
        if (observation.wirelessPolicyBlocked()) {
            return decision(Action.POLICY_BLOCKED, "Wireless Debugging disabled by device/admin policy");
        }
        if (observation.previousPairingKnown() && !observation.reconnectAttempted()) {
            return decision(Action.TRY_ADB_RECONNECT, "reuse previous pairing before opening Settings");
        }
        if (observation.screen() == Screen.CREDENTIAL_PROMPT) {
            return decision(Action.WAIT_USER_CREDENTIAL, "never automate device credential");
        }
        if (observation.screen() == Screen.PAIRING_DIALOG) {
            if (!safeObservation(observation)) {
                return decision(Action.WAIT_STABLE, "pairing dialog not stable/trusted");
            }
            var codes = observation.pairCodeCandidates().stream()
                    .filter(code -> code.length() == 6 && code.chars().allMatch(Character::isDigit))
                    .toList();
            if (observation.requestActive() && observation.pairingContextHigh() && codes.size() == 1) {
                return new Decision(Action.AUTO_PAIR_WITH_CODE, codes.get(0), "unique contextual code");
            }
            return decision(Action.SHOW_MANUAL_PAIR_CODE, "PIN not unambiguous");
        }
        if (observation.screen() == Screen.NETWORK_CONFIRMATION) {
            return safeClick(observation, "network_confirm_positive")
                    ? new Decision(Action.ACCEPT_NETWORK_CONFIRMATION,
                            "network_confirm_positive", "exact trusted confirmation")
                    : decision(Action.FALLBACK_GUIDE, "unsafe confirmation");
        }
        if (observation.screen() == Screen.WIRELESS_DEBUGGING) {
            if (Boolean.FALSE.equals(observation.wirelessEnabled())) {
                return safeClick(observation, "wireless_debugging_toggle")
                        ? new Decision(Action.ENABLE_WIRELESS_DEBUGGING,
                                "wireless_debugging_toggle", "wireless is off")
                        : decision(Action.FALLBACK_GUIDE, "unsafe toggle");
            }
            if (Boolean.TRUE.equals(observation.wirelessEnabled())) {
                return safeClick(observation, "pair_with_code")
                        ? new Decision(Action.CLICK_PAIR_WITH_CODE,
                                "pair_with_code", "wireless already enabled")
                        : decision(Action.FALLBACK_GUIDE, "unsafe pair target");
            }
            return decision(Action.WAIT_STABLE, "wireless state unknown");
        }
        if (observation.screen() == Screen.DEVELOPER_OPTIONS) {
            return safeClick(observation, "wireless_debugging")
                    ? new Decision(Action.CLICK_WIRELESS_DEBUGGING,
                            "wireless_debugging", "developer options available")
                    : decision(Action.FALLBACK_GUIDE, "wireless row unsafe");
        }
        if (observation.screen() == Screen.SOFTWARE_INFO) {
            if (observation.buildTapsDone() < 7) {
                return safeClick(observation, "build_number")
                        ? new Decision(Action.CLICK_BUILD_NUMBER, "build_number",
                                "tap " + (observation.buildTapsDone() + 1) + "/7")
                        : decision(Action.FALLBACK_GUIDE, "build number unsafe");
            }
            return decision(Action.OPEN_DEVELOPER_SETTINGS, "re-probe after seven taps");
        }
        if (observation.screen() == Screen.ABOUT_PHONE) {
            return safeClick(observation, "software_info")
                    ? new Decision(Action.CLICK_SOFTWARE_INFO, "software_info", "Samsung recipe")
                    : decision(Action.FALLBACK_GUIDE, "software info unsafe");
        }
        if (!observation.directDevProbeAttempted()) {
            return decision(Action.OPEN_DEVELOPER_SETTINGS, "fast path");
        }
        if (!observation.directDevScreenRecognized()) {
            return observation.oem().equalsIgnoreCase("Samsung")
                    ? decision(Action.OPEN_DEVICE_INFO_SETTINGS, "Samsung enable-development fallback")
                    : decision(Action.FALLBACK_GUIDE, "OEM recipe unavailable");
        }
        return decision(Action.FALLBACK_GUIDE, "unrecognized state");
    }

    private boolean safeObservation(Observation observation) {
        return observation.authority() != null && observation.authority().safe();
    }

    private boolean safeClick(Observation observation, String key) {
        return safeObservation(observation)
                && observation.candidate() != null
                && key.equals(observation.candidate().key())
                && observation.candidate().safeToClick();
    }

    private Decision decision(Action action, String reason) {
        return new Decision(action, null, reason);
    }
}

File: docs/artifacts/remote-install-autopilot/reference/AdaptiveAutopilotPlanner.kt
package reference.autopilot

enum class Confidence { LOW, MEDIUM, HIGH }
enum class Screen { UNKNOWN, APP, SETTINGS_HOME, ABOUT_PHONE, SOFTWARE_INFO, DEVELOPER_OPTIONS, WIRELESS_DEBUGGING, NETWORK_CONFIRMATION, PAIRING_DIALOG, CREDENTIAL_PROMPT }
enum class Action { WAIT_STABLE, ASK_ALLOW_RESTRICTED_SETTINGS, ASK_ENABLE_ACCESSIBILITY, ASK_CONNECT_WIFI, TRY_ADB_RECONNECT, POLICY_BLOCKED, UNSUPPORTED_ANDROID, OPEN_DEVELOPER_SETTINGS, OPEN_DEVICE_INFO_SETTINGS, CLICK_SOFTWARE_INFO, CLICK_BUILD_NUMBER, WAIT_USER_CREDENTIAL, CLICK_WIRELESS_DEBUGGING, ENABLE_WIRELESS_DEBUGGING, ACCEPT_NETWORK_CONFIRMATION, CLICK_PAIR_WITH_CODE, AUTO_PAIR_WITH_CODE, SHOW_MANUAL_PAIR_CODE, CONNECT_SUPPORT, FALLBACK_GUIDE, DONE }

data class SnapshotAuthority(
    val trustedSettingsWindow: Boolean,
    val stableSnapshots: Int,
    val generationCurrent: Boolean,
    val windowIdCurrent: Boolean,
    val fingerprintCurrent: Boolean,
    val ambiguousWindow: Boolean = false,
) {
    fun safe(): Boolean = trustedSettingsWindow && stableSnapshots >= 2 && generationCurrent && windowIdCurrent && fingerprintCurrent && !ambiguousWindow
}

data class Candidate(
    val key: String,
    val confidence: Confidence,
    val clickable: Boolean,
    val unique: Boolean,
    val marginOk: Boolean,
    val freshReacquired: Boolean,
) {
    fun safeToClick(): Boolean = confidence == Confidence.HIGH && clickable && unique && marginOk && freshReacquired
}

data class Observation(
    val androidApi: Int,
    val oem: String,
    val accessibilityEnabled: Boolean,
    val adbConnected: Boolean = false,
    val supportConnected: Boolean = false,
    val restrictedSettingsRequired: Boolean = false,
    val wifiReady: Boolean = true,
    val wirelessPolicyBlocked: Boolean = false,
    val previousPairingKnown: Boolean = false,
    val reconnectAttempted: Boolean = false,
    val screen: Screen = Screen.UNKNOWN,
    val authority: SnapshotAuthority? = null,
    val candidate: Candidate? = null,
    val wirelessEnabled: Boolean? = null,
    val buildTapsDone: Int = 0,
    val pairCodeCandidates: List<String> = emptyList(),
    val pairingContextHigh: Boolean = false,
    val requestActive: Boolean = true,
    val directDevProbeAttempted: Boolean = false,
    val directDevScreenRecognized: Boolean = false,
)

data class Decision(val action: Action, val target: String? = null, val reason: String = "")

class AdaptiveAutopilotPlanner {
    companion object { const val MIN_WIRELESS_ADB_API = 30 }

    fun decide(o: Observation): Decision {
        if (o.supportConnected) return Decision(Action.DONE, reason = "support already connected")
        if (o.adbConnected) return Decision(Action.CONNECT_SUPPORT, reason = "ADB already valid; skip Settings/pairing")
        if (o.androidApi < MIN_WIRELESS_ADB_API) return Decision(Action.UNSUPPORTED_ANDROID, reason = "Android <11 standard wireless path unsupported")
        if (!o.accessibilityEnabled && o.restrictedSettingsRequired) return Decision(Action.ASK_ALLOW_RESTRICTED_SETTINGS, reason = "sideloaded Accessibility is restricted by OS")
        if (!o.accessibilityEnabled) return Decision(Action.ASK_ENABLE_ACCESSIBILITY, reason = "manual bootstrap prerequisite")
        if (!o.wifiReady) return Decision(Action.ASK_CONNECT_WIFI, reason = "Wireless Debugging needs a usable Wi-Fi network")
        if (o.wirelessPolicyBlocked) return Decision(Action.POLICY_BLOCKED, reason = "Wireless Debugging disabled by device/admin policy")
        if (o.previousPairingKnown && !o.reconnectAttempted) return Decision(Action.TRY_ADB_RECONNECT, reason = "reuse previous pairing before opening Settings")
        if (o.screen == Screen.CREDENTIAL_PROMPT) return Decision(Action.WAIT_USER_CREDENTIAL, reason = "never automate device credential")

        if (o.screen == Screen.PAIRING_DIALOG) {
            if (!safeObservation(o)) return Decision(Action.WAIT_STABLE, reason = "pairing dialog not stable/trusted")
            val codes = o.pairCodeCandidates.filter { it.length == 6 && it.all(Char::isDigit) }
            if (o.requestActive && o.pairingContextHigh && codes.size == 1) {
                return Decision(Action.AUTO_PAIR_WITH_CODE, target = codes.single(), reason = "unique contextual code")
            }
            return Decision(Action.SHOW_MANUAL_PAIR_CODE, reason = "PIN not unambiguous")
        }

        if (o.screen == Screen.NETWORK_CONFIRMATION) {
            return if (safeClick(o, "network_confirm_positive")) Decision(Action.ACCEPT_NETWORK_CONFIRMATION, "network_confirm_positive")
            else Decision(Action.FALLBACK_GUIDE, reason = "unsafe confirmation")
        }

        if (o.screen == Screen.WIRELESS_DEBUGGING) {
            return when (o.wirelessEnabled) {
                false -> if (safeClick(o, "wireless_debugging_toggle")) Decision(Action.ENABLE_WIRELESS_DEBUGGING, "wireless_debugging_toggle") else Decision(Action.FALLBACK_GUIDE, reason = "unsafe toggle")
                true -> if (safeClick(o, "pair_with_code")) Decision(Action.CLICK_PAIR_WITH_CODE, "pair_with_code") else Decision(Action.FALLBACK_GUIDE, reason = "unsafe pair target")
                null -> Decision(Action.WAIT_STABLE, reason = "wireless state unknown")
            }
        }

        if (o.screen == Screen.DEVELOPER_OPTIONS) {
            return if (safeClick(o, "wireless_debugging")) Decision(Action.CLICK_WIRELESS_DEBUGGING, "wireless_debugging")
            else Decision(Action.FALLBACK_GUIDE, reason = "wireless row unsafe")
        }

        if (o.screen == Screen.SOFTWARE_INFO) {
            if (o.buildTapsDone < 7) {
                return if (safeClick(o, "build_number")) Decision(Action.CLICK_BUILD_NUMBER, "build_number", "tap ${o.buildTapsDone + 1}/7")
                else Decision(Action.FALLBACK_GUIDE, reason = "build number unsafe")
            }
            return Decision(Action.OPEN_DEVELOPER_SETTINGS, reason = "re-probe after seven taps")
        }

        if (o.screen == Screen.ABOUT_PHONE) {
            return if (safeClick(o, "software_info")) Decision(Action.CLICK_SOFTWARE_INFO, "software_info")
            else Decision(Action.FALLBACK_GUIDE, reason = "software info unsafe")
        }

        if (!o.directDevProbeAttempted) return Decision(Action.OPEN_DEVELOPER_SETTINGS, reason = "fast path")

        if (o.directDevProbeAttempted && !o.directDevScreenRecognized) {
            if (o.oem.equals("Samsung", ignoreCase = true)) return Decision(Action.OPEN_DEVICE_INFO_SETTINGS, reason = "Samsung enable-development fallback")
            return Decision(Action.FALLBACK_GUIDE, reason = "OEM recipe unavailable")
        }

        return Decision(Action.FALLBACK_GUIDE, reason = "unrecognized state")
    }

    private fun safeObservation(o: Observation): Boolean = o.authority?.safe() == true
    private fun safeClick(o: Observation, key: String): Boolean = safeObservation(o) && o.candidate?.let { it.key == key && it.safeToClick() } == true
}
